medical_prescription_state_verify/models/medical_prescription_order.py

```diff
@@ -2,61 +2,70 @@
-    """
-
     _inherit = 'medical.prescription.order'
 
-    _ALLOWED_CHANGE_KEYS = ['stage_id', ]
-    _ALLOWED_CHANGE_STATES = ['Verified', 'Cancelled', 'Exception', ]
-
     @api.multi
-    def write(self, vals, ):
+    def write(self, vals):
         """
-        Overload write & perform audit validations
-        Raises:
-            ValidationError: When a write is not allowed due to being in a
-                protected state
+        1) Limit changes to verified orders
+        2) Move order lines to a hold state when their orders are verified
+        3) Move order lines to an exception state when their orders are
+        cancelled or marked as exceptions
         """
-        for rec_id in self:
-            if rec_id.stage_id.is_verified:
-                # Only allow changes for keys in self._ALLOWED_CHANGE_KEYS
-                keys = vals.keys()
-                for allowed_key in self._ALLOWED_CHANGE_KEYS:
-                    try:
-                        del keys[keys.index(allowed_key)]
-                    except ValueError:
-                        pass
-                if len(keys) > 0:
-                    raise exceptions.ValidationError(_(
-                        'You cannot edit this value after an Rx has been'
-                        ' verified. Please either cancel it, or mark it as an'
-                        ' exception if manual reversals are required. [%s]' %
-                        rec_id.name
-                    ))
-
-                # Only allow state changes from self._ALLOWED_CHANGE_STATES
-                if vals.get('stage_id'):
-                    stage_id = self.env['base.kanban.stage'].browse(
-                        vals['stage_id']
+        STATE_MODULE = 'medical_prescription_state'
+        cancelled_state = self.env.ref(
+            STATE_MODULE + '.prescription_order_state_cancelled'
+        )
+        exception_state = self.env.ref(
+            STATE_MODULE + '.prescription_order_state_exception'
+        )
+        post_verify_states = cancelled_state + exception_state
+
+        verified_recs = self.filtered(lambda r: r.stage_id.is_verified)
+        if verified_recs:
+            for key in vals:
+                if key != 'stage_id':
+                    raise ValidationError(
+                        _('You are trying to edit one or more prescriptions'
+                          ' that have already been verified (e.g. %s). Please'
+                          ' either cancel them or mark them as exceptions if'
+                          ' manual changes are required.')
+                        % verified_recs[0].name
                     )
-                    if stage_id.name not in self._ALLOWED_CHANGE_STATES:
-                        raise exceptions.ValidationError(_(
-                            'You cannot move an Rx into this state after it'
-                            ' has been verified. [%s]' % rec_id.name
-                        ))
+                else:
```

Comment:
You can actually remove this else and eliminate a tad of complexity. The raise is like a return - code execution stops at that point, thus the else is superfluous